drawingBufferColorSpace support in WebGL canvas on macOS

squelart opened the pull request:
#### 0dfb608af54f0e352c7ba6879c72b1847886f585
<pre>
drawingBufferColorSpace support in WebGL canvas on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=247165">https://bugs.webkit.org/show_bug.cgi?id=247165</a>
&lt;rdar://101663516&gt;

Reviewed by Kimmo Kinnunen.

Add &apos;drawingBufferColorSpace&apos; attribute in canvas.getContext(&apos;webgl&apos; or &apos;webgl2&apos;).
The default value is &apos;srgb&apos;.
On macOS, the supported colorspaces are &apos;srgb&apos; and &apos;display-p3&apos;. Note that currently this only affects how the buffer is interpreted when presented on the screen.
On other platforms, the attribute will not even be present.

* Source/WebCore/platform/graphics/GraphicsContextGL.h:

Canonical link: <a href="https://commits.webkit.org/256208@main">https://commits.webkit.org/256208@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe11ad0941bf91f5a5f11b4e6a0e78a05c34e210

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94888 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4037 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104494 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164757 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98883 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4122 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32218 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87191 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100415 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100557 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2979 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81581 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29963 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84895 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/84438 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72850 "Found 1 new API test failure: /TestWebKit:WebKit.NewFirstVisuallyNonEmptyLayoutFrames (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38629 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18271 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36462 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19551 "Found 30 new test failures: editing/style/apply-style-join-child-text-nodes-crash.html, fast/events/wheelevent-in-frame.html, fast/frames/frame-limit.html, fast/loader/navigate-with-new-target-after-back-forward-navigation.html, http/tests/appcache/fail-on-update-2.html, http/tests/cookies/same-site/popup-same-site-post.html, http/tests/resourceLoadStatistics/third-party-cookie-blocking-ephemeral.html, http/wpt/cache-storage/cache-quota.any.html, http/wpt/fetch/response-opaque-clone.html, imported/w3c/web-platform-tests/IndexedDB/idb-explicit-commit.any.worker.html ... (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4291 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40386 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42362 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38784 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
```diff
@@ -183,6 +184,9 @@ class WebGLRenderingContextBase : public GraphicsContextGL::Client, public GPUBa
     int drawingBufferHeight() const;
 
+    PredefinedColorSpace drawingBufferColorSpace() const;
+    void setDrawingBufferColorSpace(PredefinedColorSpace colorSpace);
```
Nit: in WebKit we tend to omit argument names when they don't add anything over the type. So for this one I'd drop the `colorSpace`.

heycam reviewed (commented):
I think @kkinnunen-apple should review this.

On macOS right now, we don't handle WebGL rendering in the GPU process, but on iOS we do, so for this to work, you'll need to implement `setDrawingBufferColorSpaceIfSupported` on `RemoteGraphicsContextGLProxy` and add a new IPC message for that.

djg commented on LayoutTests/fast/canvas/webgl/colorspaces-test.html:
```diff
@@ -0,0 +1,106 @@
+<html>
+<head>
+<meta charset="utf-8">
+<title>WebGL Canvas Conformance Tests</title>
```
Perhaps change the title if this doesn't originate from the WebGL conformance test suite to avoid confusion.

djg commented on LayoutTests/fast/canvas/webgl/match-page-color-space.html:
```diff
@@ -35,7 +44,17 @@
+        if (matches) {
+            drawColorIntoCanvas([matches[1] / 255, matches[2] / 255, matches[3] / 255], canvas);
+        } else {
+            matches = /color\(([^\s]+)\s(1|0(?:\.\d+)?)\s(1|0(?:\.\d+)?)\s(1|0(?:\.\d+)?)\)/.exec(backgroundColor);
```
Wow

djg commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -1193,9 +1193,11 @@ PredefinedColorSpace WebGLRenderingContextBase::drawingBufferColorSpace() const
+        return;
+
+    if (m_drawingBufferColorSpace != colorSpace && m_context->setDrawingBufferColorSpaceIfSupported(toDestinationColorSpace(colorSpace))) {
         m_drawingBufferColorSpace = colorSpace;
```
I'm not a fan of having two pieces of state that need to be kept in sync. Is it possible to have GraphicsContextGL be the source of truth and remove the state mirroring from `WebGLRenderingContextBase`?

djg commented on Source/WebCore/platform/graphics/GraphicsContextGL.h:
```diff
@@ -1450,6 +1451,8 @@ class GraphicsContextGL : public RefCounted<GraphicsContextGL> {
 
     virtual void setContextVisibility(bool) = 0;
 
+    virtual bool setDrawingBufferColorSpaceIfSupported(DestinationColorSpace) = 0;
```
If `GraphicsContextGL` is the source of truth, then this doesn't need to return `bool`.

squelart replied:
True, but I'll then need another function to get the maybe-currently-set colorspace. But it does sound better to have one source of truth.

squelart replied:
On second thought, it would mean that getting the current attribute value from JS would require a sync message (for platforms with a GPU process).
Though in this case it may be acceptable because reads should be rare.

kkinnunen-apple replied:
you should move the function upwards along with some function that does similar context state manipulation. When it is declared in the correct block of code, you can run the generate-gpup-webgl script, and it will generate the expected functions for passing it to GPUP.

djg commented on Source/WebCore/platform/graphics/gbm/GraphicsContextGLGBM.cpp:
```diff
@@ -83,6 +83,11 @@ void GraphicsContextGLGBM::setContextVisibility(bool)
+{
+    return false;
+}
+
```
Why implement these in each derived class from `GraphicsContextGL` instead of this being the default behaviour? Less code is better code.

squelart replied:
I did try that first! But I got linker errors somewhere because it couldn't find that base-class implementation! I'll try again...

kkinnunen-apple replied:
The base declaration needs to be declared as  exported.

squelart replied:
Oh, thanks for the tip!
I did manage to fix the link issue by providing an override in RemoteGraphicsContextGLProxy.h (that's where the "missing" base implementation was called from). So do I still need to export this one?

kkinnunen-apple replied:
In case you have a non-inline base implementation, you need to export.
In case you have inline base implementation or pure virtual function, you must not export.

There should probably not be a shared implementation here, no backend can share anything useful except empty implementation and empty implementation is not that useful.

kkinnunen-apple commented on LayoutTests/fast/canvas/webgl/colorspaces-test.html:
```diff
@@ -0,0 +1,106 @@
+  return (a >= b - epsilon) && (a <= b + epsilon);
+}
+
+function checkPixels(gl, w, h, color, alpha) {
```
I think similar function should exist in WebGLTestUtils

squelart replied:
checkCanvasRectColor may indeed do the job.

kkinnunen-apple commented on LayoutTests/fast/canvas/webgl/colorspaces-test.html:
```diff
@@ -0,0 +1,106 @@
+  var consoleDiv = document.getElementById("console");
+  document.body.insertBefore(canvas, consoleDiv);
+
+  // This 'gl' must be global for shouldBe to work.
```
This is "common knowledge" and doesn't warrant a comment.

squelart replied:
I copied it from another file, happy to remove it here.

kkinnunen-apple commented on LayoutTests/fast/canvas/webgl/colorspaces-test.html:
```diff
@@ -0,0 +1,106 @@
+    checkPixels(gl, w, h, color, 1);
+  }
+
+  canvas.remove();
```
I think it's customary to leave the element to the canvas, so when the test fails it's easier to debug?

squelart replied:
I found that there's a limit of 15 webgl canvas, so I was trying not to keep them all around. But I think the test would still work even if old canvases become inactive, the text results should stay the same.

kkinnunen-apple reviewed (commented):
For the parts of the code that "there is no user-serviceable parts", the code should be just written that way by default and then fixup codepaths/ifdefs added for platforms that don't support the features. 

If the switch can be logically always succeeds or fails with context lost, there should not be conditionals or such. In this scenario, the value should logically be stored client-side.

The eventual landing commit should include the GPUP parts. It's almost there. Nice.

(and the "IfSupported" probably should be removed, as this is not such a elaborate feature that would warrant such. Either the code is there and invoked or is not there and not invoked)

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -578,10 +579,23 @@ static bool needsEAGLOnMac()
 {
     ASSERT(!getInternalFramebufferSize().isEmpty());
-    auto backing = IOSurface::create(nullptr, getInternalFramebufferSize(), DestinationColorSpace::SRGB());
+    auto backing = IOSurface::create(nullptr, getInternalFramebufferSize(), m_drawingBufferColorSpace);
```
when the buffer is sent for compositing, previous buffer is recycled in some cases.
You probably need to address this in some way in `prepareForDisplay()`
E.g. 
Buffer 1: srgb
Buffer 2: p3
Buffer 3: reuse buffer 1, need to somehow address that it should be p3.

squelart replied:
I think reusable buffers are stored with their colorspace, so there shouldn't be discrepancies; but I'll double check...

kkinnunen-apple replied:
What I mean is that consider sequence of:
1. draw 1 frame with srgb
2. switch to p3
3. draw 1 frame with p3
3.1. Here we don't want the case where we pull a buffer recycled buffer from the swap chain and it ends up containing srgb

>I think reusable buffers are stored with their colorspace, so there shouldn't be discrepancies

That's the discrepancy: if we reuse a buffer that was submitted as srgb..

squelart replied:
I meant, I think the buffers are keyed on their colorspace, so when the caller requests a P3 buffer, it should get one that was created with a P3 colorspace. But I should double/triple-check!

kkinnunen-apple replied:
To expand:
Currently prepareForDisplay() checks that the resolution of the recycled buffer matches the current resolution of the context.

In future,  prepareForDisplay() should check that the resolution of the recycled buffer matches the current resolution of the context AND that the color space matches the current drawing buffer color space of the context.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -578,15 +579,33 @@ static bool needsEAGLOnMac()
     if (m_resourceOwner)
         backing->setOwnershipIdentity(m_resourceOwner);
     backing->migrateColorSpaceToProperties();
+    m_drawingBufferColorSpace = backing->colorSpace();
```
I think here you have a redundant assignment. If this was an attempt to address the comments above, the issue is the inverse of this, not this.

squelart replied:
My idea was that we first assign `m_drawingBufferColorSpace` with the requested value, but then the "backing" IOSurface could choose another colorspace based on its capabilities, and we would record that actual value.
And that's why there are separate set&get functions, because some platforms could refuse a set("display-p3") and that would be discoverable by doing a get() that would return "srgb" -- does that make sense?

> If this was an attempt to address the comments above, the issue is the inverse of this, not this.

Sorry, I'm getting a bit lost in all these comments, which comment/issue/invers are you referring to?

My goal was to get close to how the JS interface (i.e., the `drawingBufferColorSpace` attribute) works: Set a value, but it may not actually be accepted in the engine; reading back `drawingBufferColorSpace` reveals what is really set.
But now that makes me think that if we try to set an invalid colorspace, the IOSurface would not be created, but then we should restore what was there before. It's getting harder by the minute!
If it's too awkward to design something through these comments, maybe we should chat first.

kkinnunen-apple commented on Source/WebCore/platform/graphics/DestinationColorSpace.cpp:
```diff
@@ -86,6 +86,11 @@ const DestinationColorSpace& DestinationColorSpace::DisplayP3()
 }
 #endif
 
+DestinationColorSpace::DestinationColorSpace()
```
this is unfortunate and you could side-step this for now by just not having the get... function.

squelart replied:
I believe this is also needed to receive the `Set...` message in the GPU process, the generated code does `DestinationColorSpace returnValue = { };`, so we need this default constructor there.

I think you were suggesting above to maybe use another hand-crafted enum, but we'd have to use that enum in all the `GraphicsContextGL` classes, right? It seemed better for me to try and use the real `DestinationColorSpace`.

An alternative would be to modify generate-gpu-webgl, so that instead of doing `X returnValue = { };`...`returnValue = mContext->doX();`, it would instead generate one line `X returnValue = mContext->doX();` (but that program looks quite complex and scary already!)

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -1185,6 +1185,20 @@ int WebGLRenderingContextBase::drawingBufferHeight() const
+{
+    if (isContextLost())
+        return PredefinedColorSpace::SRGB;
+    return toPredefinedColorSpace(m_context->getDrawingBufferColorSpace()).value_or(PredefinedColorSpace::SRGB);
```
I think this should just return m_drawingBufferColorSpace.

E.g. chain of logic:

All current implementations support all color spaces.

-> For all current implementations, setDrawingBufferColorSpace cannot fail.
-> if setDrawingBufferColorSpace can never fail, the get.. should not exist and we should cache this client-side.
-> Currently it is a bug to enable CanvasColorSpaceEnabled without the complete implementation.
-> Since the supported extra attributes are only P3, it's pointless to support "implementation" that doesn't support P3, as it then wouldn't support anything.

squelart replied:
I got rid of `WebGLRenderingContextBase::m_drawingBufferColorSpace`! I concurred with Dan suggesting that we should only have one source of truth -- so you think we should keep the set colorspace in both places?

> For all current implementations, setDrawingBufferColorSpace cannot fail.

Hmm, that's quite a different take that what I was working with!
So (to confirm) you're saying that whatever options exist in `PredefinedColorSpace` and/or(?) `DestinationColorSpace`, we can just assume that `setDrawingBufferColorSpace` will just deal with it, and the caller shouldn't care what really happens?
And from a JS test, `if ("drawingBufferColorSpace" in gl) { gl.drawingBufferColorSpace = "display-p3"; shouldBe(gl.drawingBufferColorSpace, "display-p3"); }` should always pass?

kkinnunen-apple replied:
> I got rid of `WebGLRenderingContextBase::m_drawingBufferColorSpace`! I concurred with Dan suggesting that we should only have one source of truth -- so you think we should keep the set colorspace in both places?

I think the way the code is currently written,`WebGLRenderingContextBase::m_drawingBufferColorSpace` is the only source of truth. 

Contrast:
- User selects clear color. We put it to m_clearColor and send it to the real context.
- User selects drawing buffer color space. We put it to m_drawingBufferColorSpace and send it to the real context.

All other things being equal, they're equal.

It makes no sense that drawing buffer color buffer would be anything else than what it was asked **if it cannot be nothing else**.  And it cannot be nothing else, if we don't have code for it to be nothing else.

> Hmm, that's quite a different take that what I was working with! So (to confirm) you're saying that whatever options exist in `PredefinedColorSpace` and/or(?) `DestinationColorSpace`, we can just assume that `setDrawingBufferColorSpace` will just deal with it, and the caller shouldn't care what really happens? And from a JS test, `if ("drawingBufferColorSpace" in gl) { gl.drawingBufferColorSpace = "display-p3"; shouldBe(gl.drawingBufferColorSpace, "display-p3"); }` should always pass?

The way the code is currently written, there is no error handling at all. If the client tries to put P3, this is propagated all the way to the IOSurface as is, and currently there is nothing else that is done or can be done.

This being the case, the simplest is to just .. write it that way explicitly.

So to recap:
- The feature should be enabled only when  PREDEFINED_COLOR_SPACE_DISPLAY_P3 is on and when the setting can be used.
- When the feature is enabled, gl.drawingBufferColorSpace = "display-p3";  selects P3.
- When P3 is selected, it is assumed to be selected. Because it **is** selected.

The situation becomes different if we have a consistent codepath which says "You tried to select P3. Currently due to some particular runtime configuration the selection could not be made, we did not respect it.". In this case, we should add the complexity of "select P3. Check if the selection succeeded or failed."

If we have such a codepath, we should also test it. E.g. write a test that ensures that P3 works. Then the test also toggles the test toggle that says "even if we support P3, manipulate the environment so we don't support P3." Then ensure that during this run, the P3 selection is not honoured.

kkinnunen-apple replied:
In slack I pasted the IOSurface.mm code that deals with the color space. Currently there's no notion of "this was not supported." as far as I understand.

kkinnunen-apple commented on LayoutTests/fast/canvas/webgl/colorspaces-test.html:
```diff
@@ -0,0 +1,103 @@
+
+    gl.clearColor(color[0], color[1], color[2], 1.0);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+    gl.flush();
```
this btw doesn't need the flush. It just confuses the reader :)

kkinnunen-apple reviewed (approved):
Looks good with some change suggestions and one bug to fix.

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContextGL.h:
```diff
@@ -1450,6 +1451,9 @@ class GraphicsContextGL : public RefCounted<GraphicsContextGL> {
     virtual void setContextVisibility(bool) = 0;
 
+    virtual DestinationColorSpace getDrawingBufferColorSpace();
+    virtual void setDrawingBufferColorSpace(const DestinationColorSpace&);
```
To be consistent, you might still mark this up with `WEBCORE_EXPORT`.
For clang, the symbol must be exported when an implementation from other library does not override the base implementation. Since all WebKit library implementations override this, it's not a problem for clang.
However,
IIRC there was might have been some related problems for other compilers (gcc or msvc) where the symbol rules might be a bit different.

kkinnunen-apple commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -1185,6 +1185,19 @@ int WebGLRenderingContextBase::drawingBufferHeight() const
     return m_context->getInternalFramebufferSize().height();
 }
 
```
Now that I remember:
You could also call the `setDrawingBufferColorSpace(m_drawingBufferColorSpace)` in `initializeNewContext()` before `reshape` call.

If you want, you can add a test where we set the colorspace to p3, lose the context, restore the context and draw and expect that the color space is the correctly retained as  p3

squelart replied:
> Now that I remember: You could also call the `setDrawingBufferColorSpace(m_drawingBufferColorSpace)` in `initializeNewContext()` before `reshape` call.

I'll try, thank you.

> If you want, you can add a test where we set the colorspace to p3, lose the context, restore the context and draw and expect that the color space is the correctly retained as p3

Good idea, will do.

> > Note about the "Style" failure: It's about missing spaces between braces, but that's consistent with other functions in InspectorCanvas.cpp.
> 
> I think it may be path of least resistance to format as the style checker tells you.
> 
> For things that are style checker bugs, you can add `// NOLINT`. However, this seems like the other functions are not following the standard.

To confirm/clarify: You'd prefer I satisfy the style checker in this one spot, rather than keeping the existing style from the 53 other similar lines? (Or should I fix the style in all those other lines as well?)

And I see some mac-wk2 failures, I'll need to fix those.

kkinnunen-apple reviewed (approved):
Still good, even better!

squelart committed:
drawingBufferColorSpace support in WebGL canvas on macOS
https://bugs.webkit.org/show_bug.cgi?id=247165
<rdar://101663516>

Reviewed by Kimmo Kinnunen.

Add 'drawingBufferColorSpace' attribute in canvas.getContext('webgl' or 'webgl2').
The default value is 'srgb'.
On macOS, the supported colorspaces are 'srgb' and 'display-p3'. Note that currently this only affects how the buffer is interpreted when presented on the screen.
On other platforms, the attribute will not even be present.

* Source/WebCore/platform/graphics/GraphicsContextGL.h:

Canonical link: https://commits.webkit.org/256208@main
